operator/pkg/clustermesh/service_informer.go

```diff
@@ -0,0 +1,344 @@
+	"github.com/cilium/cilium/pkg/service/store"
+)
+
+const (
+	meshServiceKeyLabel     = "mesh.cilium.io/service-key"
+	meshServiceClusterLabel = "mesh.cilium.io/service-cluster"
+)
+
+// meshServiceInformer uses the ClusterServices to pass fake services to the controllers
+// based on the services in the mesh.
+// Those services will be named with this form `$svcName-$clusterName` and contains
+// a mcsapi source cluster label so that the underlying endpointslice will
+// be (almost) correct. We also have a meshClient that overrides some endpointslice
+// methods to fix back the service label name.
+// The selector labels are always `mesh.cilium.io/service-key` and `mesh.cilium.io/service-cluster`
+// so that the meshPodInformer can use that for his labels.
+type meshServiceInformer struct {
+	globalServiceCache *globalServiceCache
+	services           resource.Resource[*slim_corev1.Service]
+	serviceStore       resource.Store[*slim_corev1.Service]
+
+	client              k8sClient.Clientset
+	endpointSliceLister discoverylisters.EndpointSliceLister
+
+	servicesSynced bool
+	handler        cache.ResourceEventHandler
+}
+
+func newNotFoundError(message string) *errors.StatusError {
+	return &errors.StatusError{ErrStatus: metav1.Status{
+		Status:  metav1.StatusFailure,
+		Reason:  metav1.StatusReasonNotFound,
+		Message: message,
+	}}
+}
+
+func doesServiceSyncEndpointSlice(svc *slim_corev1.Service) bool {
+	value, ok := annotation.Get(svc, annotation.GlobalService)
+	if !ok || strings.ToLower(value) != "true" {
+		return false
+	}
+
+	value, ok = annotation.Get(svc, annotation.GlobalServiceSyncEndpointSlices)
+	if !ok || strings.ToLower(value) != "true" {
+		return false
+	}
+
+	return true
+}
+
+func (i *meshServiceInformer) refreshAllCluster(ctx context.Context, svc *slim_corev1.Service) error {
+	if i.handler == nil {
+		// We don't really need to return an error here as this means that the EndpointSlice controller
+		// has not started yet and the controller will resync the initial state anyway
+		return nil
+	}
+
+	if !doesServiceSyncEndpointSlice(svc) && svc.DeletionTimestamp == nil {
+		// Usually EndpointSlice get cleaned up via OwnerReference whenever
+		// the service is deleted, but here the Service still exist but should
+		// no longer sync the remote cluster EndpintSlice, so we need to make
+		// sure the existing EndpointSlice are properly deleted.
+		labelSelector := labels.Set(map[string]string{
+			meshRealServiceNameLabel: svc.Name,
+			discovery.LabelManagedBy: ciliumControllerName,
+		}).AsSelectorPreValidated()
+		endpointSlices, err := i.endpointSliceLister.EndpointSlices(svc.Namespace).List(labelSelector)
+		if err != nil {
+			return err
+		}
+		for _, endpointSlice := range endpointSlices {
+			if err = i.client.DiscoveryV1().EndpointSlices(endpointSlice.Namespace).
+				Delete(ctx, endpointSlice.Name, metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+				return err
+			}
+		}
```

Comment:
There's the possibility of a race condition here causing leftover endpointslices if the informer doesn't yet have cached one of them (which maybe has been created just now). Could we let the epslicecontroller taking care of it by making the pod lister not return any pod if the service is not shared, and just emit a service event here to trigger reconciliation?

Reply:
Never mind, it would not work because the controller [just returns](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/vendor/github.com/cilium/endpointslice-controller/endpointslice/endpointslice_controller.go#L367) if the service does not exist. We could probably still have a service with zero endpoints, but that would leave a [placeholder epslice](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/vendor/github.com/cilium/endpointslice/reconciler.go#L231-L232) behind. Let me see if I can figure out any better alternatives.

Reply:
So for this not to be racy my proposal would be to do a list call with the client rather than the cache with ResourceVersion=0 so that we bypass all the caches. This would be a really expensive operation but I think this case should not happen often so I think it would not be worth it to try to optimize it much.

And then at this point it is still racy because when we are in that case we could still be in the middle of a reconciliation, so we need to add an extra check in the Create EndpointSlice overridden client function. Hopefully in a similar-ish case the client returns an error when there he has a OwnerReference that doesn't exist and I could return a similar error here, I need to check that.

I believe with those two things we should avoid most race condition :crossed_fingers:.

Reply:
> This would be a really expensive operation but I think this case should not happen often so I think it would not be worth it to try to optimize it much.

I'm not that sure this would hold true, unless carefully optimized to keep track of whether custom endpointslices for the same service had been created in the past (taking also into account the case of restart). 

At this point, I wonder if it wouldn't be easier to have a separate controller \[*\] which watches services (through the fake informer) and endpointslices, enqueues the key for the corresponding service, and during reconciliation checks if the service does no longer exist and in that case triggers the deletion of the remaining epslices. That would impose no additional overhead on k8s as the same informers are already active, and issue a deletion only for non-terminating epslices.

\[*\] If we are fine with modifying the upstream logic, it would be even easier to just implement it in the endpointslice controller before returning if [the service is not found](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/vendor/github.com/cilium/endpointslice-controller/endpointslice/endpointslice_controller.go#L366-L367), as we would not need a separate controller.

Reply:
> > This would be a really expensive operation but I think this case should not happen often so I think it would not be worth it to try to optimize it much.

> I'm not that sure this would hold true, unless carefully optimized to keep track of whether custom endpointslices for the same service had been created in the past (taking also into account the case of restart).

Just to clarify this case is only when you edit an existing global service that sync endpointslice and remove either of those two annotations, I think this doesn't happen very often usually I would say that creation or deletion would be way more common (and for deletion this is handled through the normal owner reference mechanism and we don't have to do anything).

So like IMO this is important for coherence to not be racy but I suspect that having this really optimized is not super important considering the very low occurrence of this specific operation (well I mean if the solution for this to be optimized is significantly more complicated).

> At this point, I wonder if it wouldn't be easier to have a separate controller [*] which watches services (through the fake informer) and endpointslices, enqueues the key for the corresponding service, and during reconciliation checks if the service does no longer exist and in that case triggers the deletion of the remaining epslices. That would impose no additional overhead on k8s as the same informers are already active, and issue a deletion only for non-terminating epslices.
 
> [*] If we are fine with modifying the upstream logic, it would be even easier to just implement it in the endpointslice controller before returning if [the service is not found](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/vendor/github.com/cilium/endpointslice-controller/endpointslice/endpointslice_controller.go#L366-L367), as we would not need a separate controller.

Wait but maybe I misunderstood this suggestion but the current code look very much what is currently implemented? The only thing missing is watching the endpointslice and executing a similar deletion code path as the current one right?

Reply:
> Just to clarify this case is only when you edit an existing global service that sync endpointslice and remove either of those two annotations

Yeah, this is what I meant above with *carefully optimized to keep track of whether custom endpointslices for the same service had been created in the past*. The current logic would trigger it for every update/deletion event of every service which doesn't have the annotations, because [`refreshAllClusters` is called for every event ](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/pkg/clustermesh/service_informer.go#L276-L280), and all non-global services would enter into the special case [here](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/pkg/clustermesh/service_informer.go#L88).

The deletion of a remote service would also cause the deletion of the corresponding epslices, because it [triggers an event through the pod informer](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/pkg/clustermesh/services.go#L57), but the reconciliation is then aborted early because the subsequent service.Get() would return not found.

> Wait but maybe I misunderstood this suggestion but the current code look very much what is currently implemented? The only thing missing is watching the endpointslice and executing a similar deletion code path as the current one right?

Yes, that is key to get notified if a new epslice gets created afterwards, and be able to remove it. The other difference would be enqueuing both updates for the real service informer (for all clusters remote clusters associated with that service) and for global service deletions, to additionally account for that case.

Reply:
> Yeah, this is what I meant above with carefully optimized to keep track of whether custom endpointslices for the same service had been created in the past. The current logic would trigger it for every update/deletion event of every service which doesn't have the annotations, because [refreshAllClusters is called for every event ](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/pkg/clustermesh/service_informer.go#L276-L280), and all non-global services would enter into the special case [here](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/pkg/clustermesh/service_informer.go#L88).

Ohhhh yes indeed I completely forgot that it was calling this for every services my bad.

So I think we should indeed just update the controller code as you were proposing at the end of your message: https://github.com/cilium/cilium/pull/28440#discussion_r1499317516. It should be a somewhat low number of LOC I believe. This file was already a bit updated for the Cilium use case, so I think that another small tweak like that (probably behind a feature flag there) should be acceptable considering the alternative!

Reply:
Sounds good to me, thanks!

Reply:
Here is the PR to handle this: https://github.com/cilium/endpointslice-controller/pull/1
+ the lastest wip in this commit that I will squash that implements this new hook

cc @aanm